Added PREF64 (#11)
ietf-wg-masque/draft-ietf-masque-connect-ip-dns: draft-ietf-masque-connect-ip-dns.md

{{?IKEv2=RFC7296}} supports discovery of DNS servers).

This document defines extensions that allow CONNECT-IP peers to convey DNS
and PREF64 configuration information to its peer using HTTP capsules
({{!HTTP-DGRAM=RFC9297}}). These extensions do not define any new way to
transport DNS queries or responses, but instead enable the exchange of DNS
resolver configuration and NAT64 prefix information inline with CONNECT-IP
sessions.

Note that these extensions are meant for cases where CONNECT-IP operates as a
Remote Access VPN (see {{Section 8.1 of CONNECT-IP}}) or a Site-to-Site VPN
(see {{Section 8.2 of CONNECT-IP}}), but not for cases like IP Flow Forwarding
(see {{Section 8.3 of CONNECT-IP}}).

For DNS configuration, this specification uses Service Bindings ({{!SVCB=RFC9460}})
to exchange information about nameservers, such as which encrypted DNS transport is
supported. This allows support for DNS over HTTPS ({{!DoH=RFC8484}}), DNS over
QUIC ({{!DoQ=RFC9250}}), DNS over TLS ({{!DoT=RFC7858}}), unencrypted DNS over
UDP port 53 and TCP port 53 ({{!DNS=RFC1035}}), as well as potential future DNS
transports. PREF64 configuration is represented in a way similar to Router

> Advertisement option defined in {{Section 4 of !PREF64-RA=RFC8781}}.

Reviewer: nit since this is just an explanation and not a dependency, it should be an informative reference

```suggestion
Advertisement option defined in {{Section 4 of ?PREF64-RA=RFC8781}}.
```

Author: Done

Similar to how Proxying IP in HTTP exchanges IP address configuration

> information ({{Section 4.7 of CONNECT-IP}}), mechanism defined in this document

Reviewer: ```suggestion
information ({{Section 4.7 of CONNECT-IP}}), the mechanism defined in this document
```

~~~
{: #ex-split title="Split Tunnel Example"}

# PREF64 Configuration Mechanism

This document defines a new capsule type, `PREF64`, that allows a CONNECT-IP peer
to communicate the IPv6 NAT64 prefix to be used for IPv6/IPv4 address synthesis.
This information enables an endpoint operating in an IPv6-only environment to
construct IPv4-reachable addresses from IPv6 literals when a NAT64 translator is in use.

## PREF64 Capsule

> Each PREF64 capsule conveys exactly one NAT64 prefix. If multiple capsules are sent

Reviewer: Since 8781 allows sending multiple PREF64s in an RA, should we also allow sending multiple prefixes per capsule?

Author: I don't think this makes sense since CONNECT-IP is naturally point-to-point. As far as I understand multiple PREF64s are possible because:
- one can have multiple gateways on the broadcast domain
- one can give them different lifetimes

Both of these points are not applicable to CONNECT-IP

Reviewer: As far as I can find from looking through Section 3 of RFC 7050, it's possible to have multiple NAT64 prefixes, and nodes are instructed to try all of them. So I assume that the motivation is to have multiple of them for redundancy. That's applicable to connect-ip. I don't think this has anything to do with broadcast domains, the NAT64s can be multiple IP hops away.

Author: Fair enough. NAT64 gateway does not have to be in the CONNECT-IP proxy, it can very much be behind it. I'll update capsule design and relevant text.

Author: Updated PR with a fresh commit. Please take another look.

~~~
{: #ex-split title="Split Tunnel Example"}

# PREF64 Configuration Mechanism

This document defines a new capsule type, `PREF64`, that allows a CONNECT-IP peer
to communicate the IPv6 NAT64 prefixes to be used for IPv6/IPv4 address synthesis.
This information enables an endpoint operating in an IPv6-only environment to
construct IPv4-reachable addresses from IPv6 literals when a NAT64 translator is in use.

## PREF64 Capsule

Each PREF64 capsule conveys zero or more NAT64 prefixes. If multiple capsules are sent
in the same direction, the most recent one replaces any previously advertised prefixes.
Empty PREF64 capsule is used to inform that NAT64 prefixes are not available.

The capsule has the following structure (see {{iana}} for the value of the capsule type):

~~~
PREF64 Capsule {
  Type (i) = PREF64,
  Length (i),
  NAT64 Prefix (104) ...,
}
~~~
{: #pref64-format title="PREF64 Capsule Format"}

> Individual NAT64 prefix has the following structure

Reviewer: ```suggestion
Individual NAT64 prefix has the following structure:
```

~~~
NAT64 Prefix {
  Prefix Length (8),
  Prefix (96),
}
~~~
{: #nat64-prefix title="NAT64 Prefix Format"}

NAT64 prefix fields are:

Prefix Length:

: The length of the NAT64 prefix in bits encoded as a single byte. Valid values are 32, 40, 48,
56, 64, and 96. These lengths correspond to the prefix sizes permitted for the IPv6-to-IPv4
address synthesis algorithm described in {{Section 2.2 of !IPv6-TRANSLATOR=RFC6052}}.

Prefix:

> : The highest 96 bits of the IPv6 prefix, encoded in network byte order.

Reviewer: ```suggestion
: The highest 96 bits of the IPv6 prefix, encoded in network byte order. Note that
this field is always 96 bits long, regardless of the value in the Prefix Length field
preceding it. The unused low order bits at the end of this field MUST be set to zero.
```

Author: I am not sure about MUST in "low order bits MUST be set to zero". RFC6052 specifically defines that lower bits are supposed to form suffix and SHOULD be set to zero. Is there another reference that mandates zero-filled suffix? If so, maybe we should make this field variable length and do not pass around meaningless zeros?

Reviewer: Oh, you're right - I actually didn't know about the [suffix](https://www.rfc-editor.org/rfc/rfc6052#section-2.2):

>   There are no remaining bits, and thus no suffix, if the prefix is 96
>   bits long.  In the other cases, the remaining bits of the address
>   constitute the suffix.  These bits are reserved for future extensions
>   and SHOULD be set to zero.  Address translators who receive IPv4-
>   embedded IPv6 addresses where these bits are not zero SHOULD ignore
>   the bits' value and proceed as if the bits' value were zero.  (Future
>   extensions may specify a different behavior.)

So, in practice those bits are almost always zero, but we shouldn't mandate they be zero. Let's just pass them through like you have it, but maybe still add a note:

```suggestion
: The highest 96 bits of the IPv6 prefix, encoded in network byte order. Note that
this field is always 96 bits long, regardless of the value in the Prefix Length field
preceding it, see {{Section 2.2 of !RFC6052}} for details.
```

Author: I've changed field name to "IPv6 Address Format" to reflect that it potentially includes both prefix and suffix and adjusted text accordingly. Please review.

Reviewer: I find this text harder to reason about. I think the previous text was fine, and matched what 8781 had. If we're unsure let's just copy 8781 and have a reference to 6052 for details.

Author: Looks like this is the right balance. Please review once again.

> ## Handling

Reviewer: Can you add a sentence like "If an endpoint receives a capsule that does not meet one of the requirements above, or with a length that is not a multiple of 13 bytes, it MUST treat it as malformed. Note that an empty PREF64 capsule is valid; it indicates that no NAT64 prefixes are available on this network."

Author: Done. Please check

~~~
PREF64 Capsule {
  Type (i) = PREF64,
  Length (i),
  NAT64 Address Format (104) ...,
}
~~~
{: #pref64-format title="PREF64 Capsule Format"}

Individual NAT64 address format has the following structure:

~~~
NAT64 Prefix {
  Prefix Length (8),
  IPv6 Address Format (96),
}
~~~
{: #nat64-prefix title="NAT64 Address Format"}

NAT64 prefix fields are:

Prefix Length:

: The length of the NAT64 prefix in bits encoded as a single byte. Valid values are 32, 40, 48,
56, 64, and 96. These lengths correspond to the prefix sizes permitted for the IPv6-to-IPv4
address synthesis algorithm described in {{Section 2.2 of !IPv6-TRANSLATOR=RFC6052}}.

IPv6 Address Format:

: Address prefix and suffix separated by 8 zero bits as defined in {{Section 2.2 of IPv6-TRANSLATOR}}.
Note that this field is always 96 bits long, regardless of the value in the Prefix Length field
preceding it. If Prefix Length is set to 96, the IPv6 Address Format consists of the prefix without a
separator or a suffix.

## Handling

Upon receiving a PREF64 capsule, a peer updates its local NAT64 configuration for the
corresponding CONNECT-IP session. The newly received PREF64 overrides any previously

> received PREF64 in the same direction.

Reviewer: ```suggestion
corresponding CONNECT-IP session. The newly received PREF64 capsule overrides any previously
received PREF64 capsules in the same direction.
```

If an endpoint receives a capsule that does not meet one of the requirements listed in {{pref64-capsule}}, or

> with a length that is not a multiple of 13 bytes, it MUST treat it as malformed. Emtpy

Reviewer: ```suggestion
with a length that is not a multiple of 13 bytes, it MUST treat it as malformed. An empty
```

# IANA Considerations {#iana}

This document, if approved, will request IANA add the following values to the
"HTTP Capsule Types" registry maintained at
<[](https://www.iana.org/assignments/masque)>.

| Value | Capsule Type |
| --- | --- |

> | 0x274C0FBB (if this document is approved, this value will be

Reviewer: Since this PR doesn't change the DNS capsule at all, should we keep the previous value of 0x1ACE79EC?
